packages/@lwc/errors/src/compiler/instrumentation.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { LWCErrorInfo } from '../shared/types';
+
+import { CompilerDiagnostic } from './utils';
+import { generateCompilerDiagnostic, ErrorConfig } from './errors';
+
+/**
+ * Pattern modeled after @lwc/engine-core's reporting.ts system
+ */
+export const enum CompilerMetrics {
+    LWCDynamicDirective = 'lwc-dynamic-directive',
+    LWCRenderModeDirective = 'lwc-render-mode-directive',
+    LWCSpreadDirective = 'lwc-spread-directive',
+    DynamicImportTransform = 'dynamic-import-transform',
+}
+
+export class CompilerInstrumentation {
+    diagnostics: CompilerDiagnostic[];
+    metrics: {
+        [key in CompilerMetrics]?: number;
```

Comment:
We can eventually introduce a payload if we feel there's more contextual information we'd like to gather that we can't get from the rest of the compilation result, but so far most of the use cases I've heard have to do with counts.